loopdb/store.go

```diff
@@ -227,59 +244,68 @@ func deserializeUpdates(swapBucket *bbolt.Bucket) ([]*LoopEvent, error) {
 	return updates, nil
 }
 
-// FetchLoopOutSwaps returns all loop out swaps currently in the store.
+// FetchLoopInSwaps returns all loop in swaps currently in the store.
 //
 // NOTE: Part of the loopdb.SwapStore interface.
-func (s *boltSwapStore) FetchLoopOutSwaps() ([]*LoopOut, error) {
-	var swaps []*LoopOut
-
-	err := s.fetchSwaps(loopOutBucketKey,
-		func(contractBytes []byte, loop Loop) error {
-			contract, err := deserializeLoopOutContract(
-				contractBytes, s.chainParams,
-			)
-			if err != nil {
-				return err
-			}
+func (s *boltSwapStore) FetchLoopInSwaps() ([]*LoopIn, error) {
+	var swaps []*LoopIn
 
-			swaps = append(swaps, &LoopOut{
-				Contract: contract,
-				Loop:     loop,
-			})
+	err := s.db.View(func(tx *bbolt.Tx) error {
+		// First, we'll grab our main loop in bucket key.
+		rootBucket := tx.Bucket(loopInBucketKey)
+		if rootBucket == nil {
+			return errors.New("bucket does not exist")
+		}
 
-			return nil
-		},
-	)
-	if err != nil {
-		return nil, err
-	}
+		// We'll now traverse the root bucket for all active swaps. The
+		// primary key is the swap hash itself.
+		return rootBucket.ForEach(func(swapHash, v []byte) error {
+			// Only go into things that we know are sub-bucket
+			// keys.
+			if v != nil {
+				return nil
+			}
 
-	return swaps, nil
-}
+			// From the root bucket, we'll grab the next swap
+			// bucket for this swap from its swaphash.
+			swapBucket := rootBucket.Bucket(swapHash)
+			if swapBucket == nil {
+				return fmt.Errorf("swap bucket %x not found",
+					swapHash)
+			}
 
-// FetchLoopInSwaps returns all loop in swaps currently in the store.
-//
-// NOTE: Part of the loopdb.SwapStore interface.
-func (s *boltSwapStore) FetchLoopInSwaps() ([]*LoopIn, error) {
-	var swaps []*LoopIn
+			// With the main swap bucket obtained, we'll grab the
+			// raw swap contract bytes and decode it.
+			contractBytes := swapBucket.Get(contractKey)
+			if contractBytes == nil {
+				return errors.New("contract not found")
+			}
 
-	err := s.fetchSwaps(loopInBucketKey,
-		func(contractBytes []byte, loop Loop) error {
 			contract, err := deserializeLoopInContract(
 				contractBytes,
 			)
 			if err != nil {
 				return err
 			}
 
-			swaps = append(swaps, &LoopIn{
+			updates, err := deserializeUpdates(swapBucket)
+			if err != nil {
+				return err
+			}
+
+			loop := LoopIn{
+				Loop: Loop{
+					Events: updates,
+				},
 				Contract: contract,
-				Loop:     loop,
-			})
+			}
+			copy(loop.Hash[:], swapHash)
```

Comment:
I think it is a copy because we don't need the validation. Otherwise it would be:
```
var err error
loop.Hash, err = lntypes.NewHash(swapHash)
if err != nil {
   return err
}
```

Reply:
I don't think it hurts to have it, so our read function's correctness is independent of the write func. But it is pre-existing, and chances of writing to disk incorrectly are slim so fine to leave as is.

Reply:
Changed to constructor